provisioning/miniprov/hare_mp/main.py

```diff
@@ -254,17 +259,148 @@ def test(args):
         exit(-1)
 
 
+def test_hare_prereq():
+    """Test suite for setting the test environment before start of test."""
+    pcs_status = ['pcs', 'status']
+    cluster_stop = ['cortx', 'cluster', 'stop']
+
+    logging.info('Test started on Host: {}'.format(execute(['hostname'])))
+    logging.info('Check that all services are up in PCS.')
+
+    resp = execute(pcs_status)
+    logging.info('PCS status: %s', resp)
+
+    if 'cluster is not currently' in resp:
+        return
+
+    logging.info('Make Node ready for testing, by stopping the cluster.')
+    resp = execute(cluster_stop)
+    logging.info('Cluster Stopped: %s', resp)
+    if 'Cluster stop is in progress' not in resp:
+        logging.error('Cluster is in progress.')
+
+
+def test_hare_postreq(cdf_file: str, timeinfo: str, logfile: str):
+    """Test suite is for restoring the setting after test."""
+    pcs_status = ['pcs', 'status']
+    cluster_start = ['cortx', 'cluster', 'start']
+
+    logging.info('Start the Cluster')
+    resp = execute(cluster_start)
+    logging.info('cluster status: %s', resp)
+    if 'Cluster start operation performed' not in resp:
+        logging.error('Cluster not yet started.')
+
+    logging.info('PCS: Check all services are up.')
+    resp = execute(pcs_status)
+    logging.info('PCS status: %s', resp)
+    if 'stopped' not in resp:
+        logging.error('Some services are not up.')
+
+    logging.info('hctl: Check that all the services are started.')
+    cluster_sts = check_cluster_status(cdf_file)
+    if cluster_sts:
+        resp = execute(['journalctl', '--since', timeinfo, '>', logfile])
+        logging.info('created journal log: %s', logfile)
+        raise Exception('hctl status reports failure.')
+
+    logging.info('Successfully performed cleanup after testing')
+
+
+def _hctlstatus():
+    """know the hctl status."""
+    hctl_status = ['hctl', 'status', '-d']
+    hctl_shutdown = ['hctl', 'shutdown', '--skip-consul-stop']
+
+    process = subprocess.Popen(hctl_status,
+                               stdin=subprocess.PIPE,
+                               stdout=subprocess.PIPE,
+                               stderr=subprocess.PIPE,
+                               encoding='utf8')
+    logging.debug('Issuing CLI command: %s', hctl_status)
+    out, err = process.communicate()
+    exit_code = process.returncode
+    logging.debug('Finished. Exit code: %d', exit_code)
+    if exit_code:
```

Comment:
This is a code duplication. Please refactor `execute()` function respectfully instead.

Reply:
hctl status returns non zero error code when cluster is not running, In this case I don't want to exit, so I added few checks.

Reply:
I don't see a problem here.
1. `execute()` raises an exception if the exit code is non-zero
2. you can catch that exception here in your code (and potentially - analyze the exception).
3. ..and keep processing after that. No exit will be triggered.

But all that unpleasant things like `PIPE`, `returncode` etc would be kept in one place only - in `execute()` function.